elastic/beats
# [Heartbeat] Add publish pipeline timeout to run_once

## What does this PR do?



Closes #35706.

Decouples run_once from sync pipeline and adds a timeout before exiting for emitting pending events.

## Checklist




## How to test this PR locally



>**Note**
> To simplify testing, it's recommended disable exiting on run_once state loader error `fmt.Errorf("run_once mode fatal error: %w", err)` 

1. Build heartbeat locally.
2. Set up a few monitors and run_once mode.
3. Point output to a non-existing ES, like `https://localhost:9200/` without it running locally.
4. Run HB, notice it never exits.
5. Stop it and add `heartbeat.publish_timeout: 15s` to `heartbeat.yml`
6. Run HB again and notice it exits after 15s, with events still pending:
```
{"log.level":"info","@timestamp":"2023-06-09T17:52:44.677+0200","log.origin":{"file.name":"beater/signalwait.go","file.line":88},"message":"Ending run_once run.","service.name":"heartbeat","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2023-06-09T17:52:44.677+0200","log.origin":{"file.name":"beater/heartbeat.go","file.line":219},"message":"Shutting down, waiting for output to complete","service.name":"heartbeat","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2023-06-09T17:52:44.677+0200","log.origin":{"file.name":"beater/heartbeat.go","file.line":230},"message":"shutdown: output timer started. Waiting for max 15s.","service.name":"heartbeat","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2023-06-09T17:52:59.681+0200","log.origin":{"file.name":"beater/signalwait.go","file.line":88},"message":"shutdown: time out waiting for pipeline to publish events.","service.name":"heartbeat","ecs.version":"1.6.0"}
```
